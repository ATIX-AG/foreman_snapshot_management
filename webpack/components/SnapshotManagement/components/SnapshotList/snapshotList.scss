
.snapshot-date {
  color: var(--pf-global--secondary-color--100);
  font-size: calc(var(--pf-global--icon--FontSize--sm) * 0.9);
}

.description {
  white-space: pre-wrap;
}

.action-buttons {
  width: 20%;
}
