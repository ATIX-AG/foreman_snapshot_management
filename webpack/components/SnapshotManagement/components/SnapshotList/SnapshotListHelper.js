import React from 'react';
import { Button, Icon } from 'patternfly-react';

import { sprintf, translate as __ } from 'foremanReact/common/I18n';

import './snapshotList.scss';

export const renderListEntryButtons = (
  canDelete,
  canRevert,
  canUpdate,
  host,
  snapshotRollbackAction,
  snapshotDeleteAction,
  inlineEditController
) => disabled => (value, additionalData) => {
  const buttons = [];

  // Edit Button
  if (canUpdate) {
    buttons.push(
      <Button
        key="edit-button"
        bsStyle="default"
        disabled={disabled}
        onClick={() => inlineEditController.onActivate(additionalData)}
      >
        <Icon type="pf" name="edit" title={__('edit entry')} />
      </Button>
    );
  }

  // Rollback button
  if (canRevert) {
    buttons.push(
      <Button
        key="rollback-button"
        bsStyle="default"
        disabled={disabled}
        onClick={() =>
          window.confirm(
            sprintf(__('Rollback to "%s"?'), additionalData.rowData.name)
          ) && snapshotRollbackAction(host, additionalData.rowData)
        }
      >
        <Icon type="pf" name="history" title={__('Rollback')} />
      </Button>
    );
  }

  // Delete Button
  if (canDelete) {
    buttons.push(
      <Button
        key="delete-button"
        bsStyle="default"
        disabled={disabled}
        onClick={() =>
          window.confirm(
            sprintf(__('Delete Snapshot "%s"?'), additionalData.rowData.name)
          ) && snapshotDeleteAction(host, additionalData.rowData)
        }
      >
        <Icon type="pf" name="delete" title={__('Delete')} />
      </Button>
    );
  }

  return <td className="action-buttons">{buttons}</td>;
};
