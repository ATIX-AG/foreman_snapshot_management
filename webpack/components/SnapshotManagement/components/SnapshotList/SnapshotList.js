import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Alert,
  // Button,
  Table as PfTable,
  FormControl,
  inlineEditFormatterFactory,
} from 'patternfly-react';
import { cloneDeep, findIndex } from 'lodash';

import { translate as __ } from 'foremanReact/common/I18n';
import {
  Table,
  TableBody,
  column,
  headerFormatterWithProps,
  cellFormatterWithProps,
} from 'foremanReact/components/common/table';
import ShortDateTime from 'foremanReact/components/common/dates/ShortDateTime';
import AlertBody from 'foremanReact/components/common/Alert/AlertBody';
import Loading from 'foremanReact/components/Loading/Loading';
import 'foremanReact/redux/API';

import { renderListEntryButtons } from './SnapshotListHelper';
import './snapshotList.scss';

class SnapshotList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      columns: this.defineColumns(),
      editMode: false,
      rows: cloneDeep(props.snapshots),
    };
  }

  defineColumns() {
    const {
      canDelete,
      canRevert,
      canUpdate,
      host,
      deleteAction,
      updateAction,
      rollbackAction,
    } = this.props;

    const inlineEditController = {
      isEditing: ({ rowData }) => rowData.backup !== undefined,
      onActivate: ({ rowData }) => {
        const rows = cloneDeep(this.state.rows);
        const index = findIndex(rows, { id: rowData.id });

        rows[index].backup = cloneDeep(rows[index]);

        this.setState({ rows, editMode: true });
      },
      onConfirm: ({ rowData }) => {
        const rows = cloneDeep(this.state.rows);
        const index = findIndex(rows, { id: rowData.id });

        delete rows[index].backup;

        this.setState({ rows, editMode: false });
        updateAction(host, rowData);
      },
      onCancel: ({ rowData }) => {
        const rows = cloneDeep(this.state.rows);
        const index = findIndex(rows, { id: rowData.id });

        rows[index] = cloneDeep(rows[index].backup);
        delete rows[index].backup;

        this.setState({ rows, editMode: false });
      },
      onChange: (value, { rowData, property }) => {
        const rows = cloneDeep(this.state.rows);
        const index = findIndex(rows, { id: rowData.id });

        rows[index][property] = value;

        this.setState({ rows });
      },
    };
    this.inlineEditController = inlineEditController;

    const renderButtons = renderListEntryButtons(
      canDelete,
      canRevert,
      canUpdate,
      host,
      rollbackAction,
      deleteAction,
      inlineEditController
    );
    const inlineEditButtonCellFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => this.state.editMode,
      renderValue: renderButtons(false),
      renderEdit: renderButtons(true),
    });

    const inlineEditFormatter = inlineEditFormatterFactory({
      isEditing: additionalData =>
        inlineEditController.isEditing(additionalData),
      renderValue: (value, additionalData) => {
        let date = '';
        if (additionalData.property === 'name')
          date = (
            <span className="snapshot-date">
              <ShortDateTime
                date={new Date(additionalData.rowData.formatted_created_at)}
                defaultValue={__('N/A')}
                showRelativeTimeTooltip
              />
            </span>
          );
        return (
          <span className="static description">
            {value}
            <br />
            {date}
          </span>
        );
      },
      renderEdit: (value, additionalData) => {
        let type = 'input';
        if (additionalData.property === 'description') type = 'textarea';
        return (
          <FormControl
            type="text"
            defaultValue={value}
            onBlur={e =>
              inlineEditController.onChange(e.target.value, additionalData)
            }
            componentClass={type}
          />
        );
      },
    });
    const editCellFormatters = [cellFormatterWithProps];
    if (canUpdate) editCellFormatters.unshift(inlineEditFormatter);

    const columns = [
      column(
        'name',
        __('Snapshot'),
        [headerFormatterWithProps],
        editCellFormatters
      ),
      column(
        'description',
        __('Description'),
        [headerFormatterWithProps],
        editCellFormatters
      ),
    ];
    if (canDelete || canUpdate || canRevert)
      columns.push(
        column(
          '',
          __('Action'),
          [headerFormatterWithProps],
          [inlineEditButtonCellFormatter],
          { className: 'action-buttons' }
        )
      );
    return columns;
  }

  componentDidMount() {
    this.props.loadSnapshots(this.props.host.id);
  }

  componentWillReceiveProps(newProps) {
    if (newProps.snapshots !== this.props.snapshots) {
      this.setState({ rows: newProps.snapshots });
    }
    if (newProps.needsReload) {
      newProps.loadSnapshots(newProps.host.id);
    }
  }

  getBodyMessage() {
    if (this.props.isLoading || this.props.isWorking)
      return <Loading textSize="sm" />;

    if (this.props.hasError)
      return (
        <Alert
          variant="danger"
          isInline
          title={__('Failed to load snapshot list')}
        >
          <AlertBody title={__('Failed to load snapshot list')}>
            {/* IMHO the line-break should be done by AlertBody :-( */}
            <br />
            {this.props.error.message}
          </AlertBody>
        </Alert>
      );

    return undefined;
  }

  render() {
    const { columns } = this.state;
    const bodyMessage = this.getBodyMessage();

    return (
      <div>
        {/*
        <Button
          disabled={this.props.isLoading}
          onClick={() => this.props.loadSnapshots(this.props.host.id)}
        >
          ReloadData
        </Button>
        */}
        <Table
          caption="Snapshot List"
          columns={columns}
          rows={this.state.rows}
          bodyMessage={bodyMessage}
          inlineEdit
          components={{
            body: {
              row: PfTable.InlineEditRow,
              cell: cellProps => cellProps.children,
            },
          }}
        >
          <PfTable.Header key="header" />
          <TableBody
            key="body"
            columns={columns}
            rows={this.state.rows}
            message={bodyMessage}
            rowKey="id"
            onRow={(rowData, { rowIndex }) => ({
              role: 'row',
              isEditing: () => this.inlineEditController.isEditing({ rowData }),
              onCancel: () =>
                this.inlineEditController.onCancel({ rowData, rowIndex }),
              onConfirm: () =>
                this.inlineEditController.onConfirm({ rowData, rowIndex }),
              last: rowIndex === this.state.rows.length - 1,
            })}
          />
        </Table>
      </div>
    );
  }
}

SnapshotList.propTypes = {
  /*
  children: PropTypes.node,
  className: PropTypes.string,
  */
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  loadSnapshots: PropTypes.func.isRequired,
  deleteAction: PropTypes.func.isRequired,
  updateAction: PropTypes.func.isRequired,
  rollbackAction: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  isWorking: PropTypes.bool,
  hasError: PropTypes.bool,
  error: PropTypes.shape({ message: PropTypes.string }),
  snapshots: PropTypes.array,

  // permissions:
  canDelete: PropTypes.bool,
  canRevert: PropTypes.bool,
  canUpdate: PropTypes.bool,
};

SnapshotList.defaultProps = {
  /*
  className: '',
  children: null,
  */
  isLoading: true,
  isWorking: false,
  hasError: false,
  error: undefined,
  snapshots: [],
  canDelete: false,
  canRevert: false,
  canUpdate: false,
};

export default SnapshotList;
