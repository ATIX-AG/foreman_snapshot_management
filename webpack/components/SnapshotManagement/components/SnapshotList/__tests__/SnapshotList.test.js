import {
  testComponentSnapshotsWithFixtures,
  // shallowRenderComponentWithFixtures,
} from 'react-redux-test-utils';
// import configureStore from 'redux-mock-store';

import SnapshotList from '../SnapshotList';

// const mockStore = configureStore([]);

const funcDummies = {
  loadSnapshots: hostId => null,
  deleteAction: hostId => null,
  rollbackAction: hostId => null,
  updateAction: hostId => null,
};
const host = {
  id: 42,
  name: 'deep.thought',
};
const testSnapshots = [
  {
    id: 'snapshot-15776',
    name: 'test',
    description: '123 testSnapshot',
    created_at: '2021-01-19 15:03:47 UTC',
  },
  {
    id: 'snapshot-17285',
    name: 'Hello',
    description: 'World\nHow are you?',
    created_at: '2021-04-28 10:22:06 UTC',
  },
];
const fixtures = {
  'without optional Props': { host, ...funcDummies },
  loading: { host, isLoading: true, ...funcDummies },
  working: { host, isWorking: true, ...funcDummies },
  error: {
    host,
    hasError: true,
    error: { message: 'TEST' },
    ...funcDummies,
  },
  'snapshot list': { host, snapshots: testSnapshots, ...funcDummies },
  'without any permissions': {
    host,
    snapshots: testSnapshots,
    canUpdate: false,
    canRevert: false,
    canDelete: false,
    ...funcDummies,
  },
  'without delete permission': {
    host,
    snapshots: testSnapshots,
    canUpdate: true,
    canRevert: true,
    canDelete: false,
    ...funcDummies,
  },
};

describe('SnapshotList', () => {
  /*
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      myState: 'sample text',
    });
    component = renderer.create(
      <Provider store={store}>
        <SnapshotList />
      </Provider>
    );
  });
  */
  describe('renders', () =>
    testComponentSnapshotsWithFixtures(SnapshotList, fixtures));
});
