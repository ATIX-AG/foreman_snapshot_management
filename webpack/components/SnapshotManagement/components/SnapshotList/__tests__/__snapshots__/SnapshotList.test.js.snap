// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotList renders error 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={Array []}
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={Array []}
    />
  </Table>
</div>
`;

exports[`SnapshotList renders loading 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={Array []}
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={Array []}
    />
  </Table>
</div>
`;

exports[`SnapshotList renders snapshot list 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={
      Array [
        Object {
          "created_at": "2021-01-19 15:03:47 UTC",
          "description": "123 testSnapshot",
          "id": "snapshot-15776",
          "name": "test",
        },
        Object {
          "created_at": "2021-04-28 10:22:06 UTC",
          "description": "World
How are you?",
          "id": "snapshot-17285",
          "name": "Hello",
        },
      ]
    }
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={
        Array [
          Object {
            "created_at": "2021-01-19 15:03:47 UTC",
            "description": "123 testSnapshot",
            "id": "snapshot-15776",
            "name": "test",
          },
          Object {
            "created_at": "2021-04-28 10:22:06 UTC",
            "description": "World
How are you?",
            "id": "snapshot-17285",
            "name": "Hello",
          },
        ]
      }
    />
  </Table>
</div>
`;

exports[`SnapshotList renders without any permissions 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={
      Array [
        Object {
          "created_at": "2021-01-19 15:03:47 UTC",
          "description": "123 testSnapshot",
          "id": "snapshot-15776",
          "name": "test",
        },
        Object {
          "created_at": "2021-04-28 10:22:06 UTC",
          "description": "World
How are you?",
          "id": "snapshot-17285",
          "name": "Hello",
        },
      ]
    }
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={
        Array [
          Object {
            "created_at": "2021-01-19 15:03:47 UTC",
            "description": "123 testSnapshot",
            "id": "snapshot-15776",
            "name": "test",
          },
          Object {
            "created_at": "2021-04-28 10:22:06 UTC",
            "description": "World
How are you?",
            "id": "snapshot-17285",
            "name": "Hello",
          },
        ]
      }
    />
  </Table>
</div>
`;

exports[`SnapshotList renders without delete permission 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Action",
            "props": Object {
              "className": "action-buttons",
            },
          },
          "property": "",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={
      Array [
        Object {
          "created_at": "2021-01-19 15:03:47 UTC",
          "description": "123 testSnapshot",
          "id": "snapshot-15776",
          "name": "test",
        },
        Object {
          "created_at": "2021-04-28 10:22:06 UTC",
          "description": "World
How are you?",
          "id": "snapshot-17285",
          "name": "Hello",
        },
      ]
    }
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Action",
              "props": Object {
                "className": "action-buttons",
              },
            },
            "property": "",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={
        Array [
          Object {
            "created_at": "2021-01-19 15:03:47 UTC",
            "description": "123 testSnapshot",
            "id": "snapshot-15776",
            "name": "test",
          },
          Object {
            "created_at": "2021-04-28 10:22:06 UTC",
            "description": "World
How are you?",
            "id": "snapshot-17285",
            "name": "Hello",
          },
        ]
      }
    />
  </Table>
</div>
`;

exports[`SnapshotList renders without optional Props 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={Array []}
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={Array []}
    />
  </Table>
</div>
`;

exports[`SnapshotList renders working 1`] = `
<div>
  <Table
    bodyMessage={
      <Loading
        iconSize="xl"
        showText={true}
        textSize="sm"
      />
    }
    caption="Snapshot List"
    columns={
      Array [
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Snapshot",
            "props": Object {},
          },
          "property": "name",
        },
        Object {
          "cell": Object {
            "formatters": Array [
              [Function],
            ],
            "props": Object {},
          },
          "header": Object {
            "formatters": Array [
              [Function],
            ],
            "label": "Description",
            "props": Object {},
          },
          "property": "description",
        },
      ]
    }
    components={
      Object {
        "body": Object {
          "cell": [Function],
          "row": [Function],
        },
      }
    }
    inlineEdit={true}
    rows={Array []}
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Snapshot",
              "props": Object {},
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {},
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Description",
              "props": Object {},
            },
            "property": "description",
          },
        ]
      }
      key="body"
      message={
        <Loading
          iconSize="xl"
          showText={true}
          textSize="sm"
        />
      }
      onRow={[Function]}
      rowKey="id"
      rows={Array []}
    />
  </Table>
</div>
`;
